// @airgap/beacon-sdk/dist/cjs/transports/clients/P2PCommunicationClient.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.P2PCommunicationClient = void 0;
var sodium = __importStar(require("libsodium-wrappers"));
var axios_1 = __importDefault(require("axios"));
var crypto_1 = require("../../utils/crypto");
var MatrixClient_1 = require("../../matrix-client/MatrixClient");
var MatrixClientEvent_1 = require("../../matrix-client/models/MatrixClientEvent");
var MatrixMessage_1 = require("../../matrix-client/models/MatrixMessage");
var __1 = require("../..");
var constants_1 = require("../../constants");
var generate_uuid_1 = require("../../utils/generate-uuid");
var get_sender_id_1 = require("../../utils/get-sender-id");
var Logger_1 = require("../../utils/Logger");
var CommunicationClient_1 = require("./CommunicationClient");
var exposed_promise_1 = require("../../utils/exposed-promise");
var logger = new Logger_1.Logger('P2PCommunicationClient');
var KNOWN_RELAY_SERVERS = [
    'beacon-node-1.sky.papers.tech',
    'beacon-node-0.papers.tech:8448',
    'beacon-node-2.sky.papers.tech'
];
var publicKeyToNumber = function (arr, mod) {
    var sum = 0;
    for (var i = 0; i < arr.length; i++) {
        sum += arr[i] + i;
    }
    return Math.floor(sum % mod);
};
/**
 * @internalapi
 *
 *
 */
var P2PCommunicationClient = /** @class */ (function (_super) {
    __extends(P2PCommunicationClient, _super);
    function P2PCommunicationClient(name, keyPair, replicationCount, storage, matrixNodes, iconUrl, appUrl) {
        var _this = _super.call(this, keyPair) || this;
        _this.name = name;
        _this.replicationCount = replicationCount;
        _this.storage = storage;
        _this.iconUrl = iconUrl;
        _this.appUrl = appUrl;
        _this.client = new exposed_promise_1.ExposedPromise();
        _this.activeListeners = new Map();
        _this.ignoredRooms = [];
        _this.loginCounter = 0;
        logger.log('constructor', 'P2PCommunicationClient created');
        _this.KNOWN_RELAY_SERVERS = matrixNodes.length > 0 ? matrixNodes : KNOWN_RELAY_SERVERS;
        return _this;
    }
    P2PCommunicationClient.prototype.getPairingRequestInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            var info, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = {};
                        return [4 /*yield*/, generate_uuid_1.generateGUID()];
                    case 1:
                        _a.id = _b.sent(),
                            _a.type = 'p2p-pairing-request',
                            _a.name = this.name,
                            _a.version = constants_1.BEACON_VERSION;
                        return [4 /*yield*/, this.getPublicKey()];
                    case 2:
                        _a.publicKey = _b.sent();
                        return [4 /*yield*/, this.getRelayServer()];
                    case 3:
                        info = (_a.relayServer = _b.sent(),
                            _a);
                        if (this.iconUrl) {
                            info.icon = this.iconUrl;
                        }
                        if (this.appUrl) {
                            info.appUrl = this.appUrl;
                        }
                        return [2 /*return*/, info];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.getPairingResponseInfo = function (request) {
        return __awaiter(this, void 0, void 0, function () {
            var info, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = {
                            id: request.id,
                            type: 'p2p-pairing-response',
                            name: this.name,
                            version: constants_1.BEACON_VERSION
                        };
                        return [4 /*yield*/, this.getPublicKey()];
                    case 1:
                        _a.publicKey = _b.sent();
                        return [4 /*yield*/, this.getRelayServer()];
                    case 2:
                        info = (_a.relayServer = _b.sent(),
                            _a);
                        if (this.iconUrl) {
                            info.icon = this.iconUrl;
                        }
                        if (this.appUrl) {
                            info.appUrl = this.appUrl;
                        }
                        return [2 /*return*/, info];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.getRelayServer = function () {
        return __awaiter(this, void 0, void 0, function () {
            var node, hasDoneMigration, preservedState, node_1, startIndex, offset, serverIndex, server, relayError_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.relayServer) {
                            return [2 /*return*/, this.relayServer.promise];
                        }
                        else {
                            this.relayServer = new exposed_promise_1.ExposedPromise();
                        }
                        return [4 /*yield*/, this.storage.get(__1.StorageKey.MATRIX_SELECTED_NODE)];
                    case 1:
                        node = _a.sent();
                        if (!(node && node.length > 0)) return [3 /*break*/, 2];
                        this.relayServer.resolve(node);
                        return [2 /*return*/, node];
                    case 2:
                        if (!(KNOWN_RELAY_SERVERS === this.KNOWN_RELAY_SERVERS)) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.storage.get(__1.StorageKey.MULTI_NODE_SETUP_DONE)];
                    case 3:
                        hasDoneMigration = _a.sent();
                        if (!!hasDoneMigration) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.storage.get(__1.StorageKey.MATRIX_PRESERVED_STATE)];
                    case 4:
                        preservedState = _a.sent();
                        console.log('PRESERVED STATE', preservedState);
                        if (preservedState.syncToken || preservedState.rooms) {
                            node_1 = 'matrix.papers.tech' // 2.2.7 Migration: This will default to the old default to avoid peers from losing their relayServer.
                            ;
                            this.storage
                                .set(__1.StorageKey.MATRIX_SELECTED_NODE, node_1)
                                .catch(function (error) { return logger.log(error); });
                            this.relayServer.resolve(node_1);
                            return [2 /*return*/, node_1];
                        }
                        this.storage.set(__1.StorageKey.MULTI_NODE_SETUP_DONE, true).catch(function (error) { return logger.log(error); });
                        _a.label = 5;
                    case 5:
                        console.log('GET RELAY SERVER');
                        startIndex = publicKeyToNumber(this.keyPair.publicKey, this.KNOWN_RELAY_SERVERS.length);
                        offset = 0;
                        _a.label = 6;
                    case 6:
                        if (!(offset < this.KNOWN_RELAY_SERVERS.length)) return [3 /*break*/, 11];
                        serverIndex = (startIndex + offset) % this.KNOWN_RELAY_SERVERS.length;
                        server = this.KNOWN_RELAY_SERVERS[serverIndex];
                        _a.label = 7;
                    case 7:
                        _a.trys.push([7, 9, , 10]);
                        return [4 /*yield*/, axios_1.default.get("https://" + server + "/_matrix/client/versions")];
                    case 8:
                        _a.sent();
                        this.storage
                            .set(__1.StorageKey.MATRIX_SELECTED_NODE, server)
                            .catch(function (error) { return logger.log(error); });
                        this.relayServer.resolve(server);
                        return [2 /*return*/, server];
                    case 9:
                        relayError_1 = _a.sent();
                        logger.log("Ignoring server \"" + server + "\", trying another one...");
                        offset++;
                        return [3 /*break*/, 10];
                    case 10: return [3 /*break*/, 6];
                    case 11:
                        this.relayServer.reject("No matrix server reachable!");
                        throw new Error("No matrix server reachable!");
                }
            });
        });
    };
    P2PCommunicationClient.prototype.tryJoinRooms = function (roomId, retry) {
        if (retry === void 0) { retry = 1; }
        return __awaiter(this, void 0, void 0, function () {
            var error_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.client.promise];
                    case 1: return [4 /*yield*/, (_a.sent()).joinRooms(roomId)];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        if (retry <= 10 && error_1.errcode === 'M_FORBIDDEN') {
                            // If we join the room too fast after receiving the invite, the server can accidentally reject our join. This seems to be a problem only when using a federated multi-node setup. Usually waiting for a couple milliseconds solves the issue, but to handle lag, we will keep retrying for 2 seconds.
                            logger.log("Retrying to join...", error_1);
                            setTimeout(function () { return __awaiter(_this, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.tryJoinRooms(roomId, retry + 1)];
                                        case 1:
                                            _a.sent();
                                            return [2 /*return*/];
                                    }
                                });
                            }); }, 200);
                        }
                        else {
                            logger.log("Failed to join after " + retry + " tries.", error_1);
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var loginRawDigest, rawSignature, relayServer, client, _a, _b, _c, _d, _e, _f, _g, error_2;
            var _this = this;
            return __generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        logger.log('start', 'starting client');
                        return [4 /*yield*/, sodium.ready];
                    case 1:
                        _h.sent();
                        loginRawDigest = sodium.crypto_generichash(32, sodium.from_string("login:" + Math.floor(Date.now() / 1000 / (5 * 60))));
                        rawSignature = sodium.crypto_sign_detached(loginRawDigest, this.keyPair.privateKey);
                        logger.log('start', "connecting to server");
                        return [4 /*yield*/, this.getRelayServer()];
                    case 2:
                        relayServer = _h.sent();
                        client = MatrixClient_1.MatrixClient.create({
                            baseUrl: "https://" + relayServer,
                            storage: this.storage
                        });
                        this.initialListener = function (event) { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                if (this.initialEvent && this.initialEvent.timestamp && event && event.timestamp) {
                                    if (this.initialEvent.timestamp < event.timestamp) {
                                        this.initialEvent = event;
                                    }
                                }
                                else {
                                    this.initialEvent = event;
                                }
                                return [2 /*return*/];
                            });
                        }); };
                        client.subscribe(MatrixClientEvent_1.MatrixClientEventType.MESSAGE, this.initialListener);
                        client.subscribe(MatrixClientEvent_1.MatrixClientEventType.INVITE, function (event) { return __awaiter(_this, void 0, void 0, function () {
                            var member;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (event.content.members.length === 1) {
                                            // If there is only one member we know it's a new room
                                            // TODO: Use the "sender" of the event instead
                                            member = event.content.members[0];
                                        }
                                        return [4 /*yield*/, this.tryJoinRooms(event.content.roomId)];
                                    case 1:
                                        _a.sent();
                                        if (!member) return [3 /*break*/, 4];
                                        return [4 /*yield*/, this.updateRelayServer(member)];
                                    case 2:
                                        _a.sent();
                                        return [4 /*yield*/, this.updatePeerRoom(member, event.content.roomId)];
                                    case 3:
                                        _a.sent();
                                        _a.label = 4;
                                    case 4: return [2 /*return*/];
                                }
                            });
                        }); });
                        _b = (_a = logger).log;
                        _c = ['start', 'login'];
                        return [4 /*yield*/, this.getPublicKeyHash()];
                    case 3:
                        _b.apply(_a, _c.concat([_h.sent(), 'on', relayServer]));
                        _h.label = 4;
                    case 4:
                        _h.trys.push([4, 8, , 10]);
                        _e = (_d = client).start;
                        _f = {};
                        return [4 /*yield*/, this.getPublicKeyHash()];
                    case 5:
                        _f.id = _h.sent();
                        _g = "ed:" + crypto_1.toHex(rawSignature) + ":";
                        return [4 /*yield*/, this.getPublicKey()];
                    case 6: return [4 /*yield*/, _e.apply(_d, [(_f.password = _g + (_h.sent()),
                                _f.deviceId = crypto_1.toHex(this.keyPair.publicKey),
                                _f)])];
                    case 7:
                        _h.sent();
                        return [3 /*break*/, 10];
                    case 8:
                        error_2 = _h.sent();
                        console.log('ERROR, RETRYING');
                        return [4 /*yield*/, this.reset()]; // If we can't log in, let's reset
                    case 9:
                        _h.sent(); // If we can't log in, let's reset
                        console.log('TRYING AGAIN');
                        if (this.loginCounter <= this.KNOWN_RELAY_SERVERS.length) {
                            this.loginCounter++;
                            this.start();
                            return [2 /*return*/];
                        }
                        else {
                            throw new Error('Too many login attempts. Try again later.');
                        }
                        return [3 /*break*/, 10];
                    case 10:
                        console.log('client is ready');
                        this.client.resolve(client);
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.stop = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.client.isResolved()) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.client.promise];
                    case 1: return [4 /*yield*/, (_a.sent()).stop().catch(function (error) { return logger.error(error); })];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [4 /*yield*/, this.reset()];
                    case 4:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.reset = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.delete(__1.StorageKey.MATRIX_PEER_ROOM_IDS).catch(function (error) { return logger.log(error); })];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.storage.delete(__1.StorageKey.MATRIX_PRESERVED_STATE).catch(function (error) { return logger.log(error); })];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.storage.delete(__1.StorageKey.MATRIX_SELECTED_NODE).catch(function (error) { return logger.log(error); })
                            // Instead of resetting everything, maybe we should make sure a new instance is created?
                        ];
                    case 3:
                        _a.sent();
                        // Instead of resetting everything, maybe we should make sure a new instance is created?
                        this.relayServer = undefined;
                        this.client = new exposed_promise_1.ExposedPromise();
                        this.initialEvent = undefined;
                        this.initialListener = undefined;
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.listenForEncryptedMessage = function (senderPublicKey, messageCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var sharedRx, callbackFunction, lastEvent, initialListener;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.activeListeners.has(senderPublicKey)) {
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, this.createCryptoBoxServer(senderPublicKey, this.keyPair.privateKey)];
                    case 1:
                        sharedRx = (_a.sent()).sharedRx;
                        callbackFunction = function (event) { return __awaiter(_this, void 0, void 0, function () {
                            var _a, payload, decryptedMessage, decryptionError_1;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        _a = this.isTextMessage(event.content);
                                        if (!_a) return [3 /*break*/, 2];
                                        return [4 /*yield*/, this.isSender(event, senderPublicKey)];
                                    case 1:
                                        _a = (_b.sent());
                                        _b.label = 2;
                                    case 2:
                                        if (!_a) return [3 /*break*/, 8];
                                        payload = void 0;
                                        return [4 /*yield*/, this.updateRelayServer(event.content.message.sender)];
                                    case 3:
                                        _b.sent();
                                        return [4 /*yield*/, this.updatePeerRoom(event.content.message.sender, event.content.roomId)];
                                    case 4:
                                        _b.sent();
                                        try {
                                            payload = Buffer.from(event.content.message.content, 'hex');
                                            // content can be non-hex if it's a connection open request
                                        }
                                        catch (_c) {
                                            /* */
                                        }
                                        if (!(payload &&
                                            payload.length >= sodium.crypto_secretbox_NONCEBYTES + sodium.crypto_secretbox_MACBYTES)) return [3 /*break*/, 8];
                                        _b.label = 5;
                                    case 5:
                                        _b.trys.push([5, 7, , 8]);
                                        return [4 /*yield*/, crypto_1.decryptCryptoboxPayload(payload, sharedRx)
                                            // logger.log(
                                            //   'listenForEncryptedMessage',
                                            //   'encrypted message received',
                                            //   decryptedMessage,
                                            //   await new Serializer().deserialize(decryptedMessage)
                                            // )
                                            // console.log('calculated sender ID', await getSenderId(senderPublicKey))
                                            // TODO: Add check for correct decryption key / sender ID
                                        ];
                                    case 6:
                                        decryptedMessage = _b.sent();
                                        // logger.log(
                                        //   'listenForEncryptedMessage',
                                        //   'encrypted message received',
                                        //   decryptedMessage,
                                        //   await new Serializer().deserialize(decryptedMessage)
                                        // )
                                        // console.log('calculated sender ID', await getSenderId(senderPublicKey))
                                        // TODO: Add check for correct decryption key / sender ID
                                        messageCallback(decryptedMessage);
                                        return [3 /*break*/, 8];
                                    case 7:
                                        decryptionError_1 = _b.sent();
                                        return [3 /*break*/, 8];
                                    case 8: return [2 /*return*/];
                                }
                            });
                        }); };
                        this.activeListeners.set(senderPublicKey, callbackFunction);
                        return [4 /*yield*/, this.client.promise];
                    case 2:
                        (_a.sent()).subscribe(MatrixClientEvent_1.MatrixClientEventType.MESSAGE, callbackFunction);
                        lastEvent = this.initialEvent;
                        if (!(lastEvent &&
                            lastEvent.timestamp &&
                            new Date().getTime() - lastEvent.timestamp < 5 * 60 * 1000)) return [3 /*break*/, 4];
                        logger.log('listenForEncryptedMessage', 'Handling previous event');
                        return [4 /*yield*/, callbackFunction(lastEvent)];
                    case 3:
                        _a.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        logger.log('listenForEncryptedMessage', 'No previous event found');
                        _a.label = 5;
                    case 5:
                        initialListener = this.initialListener;
                        if (!initialListener) return [3 /*break*/, 7];
                        ;
                        return [4 /*yield*/, this.client.promise];
                    case 6:
                        (_a.sent()).unsubscribe(MatrixClientEvent_1.MatrixClientEventType.MESSAGE, initialListener);
                        _a.label = 7;
                    case 7:
                        this.initialListener = undefined;
                        this.initialEvent = undefined;
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.unsubscribeFromEncryptedMessage = function (senderPublicKey) {
        return __awaiter(this, void 0, void 0, function () {
            var listener;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        listener = this.activeListeners.get(senderPublicKey);
                        if (!listener) {
                            return [2 /*return*/];
                        }
                        ;
                        return [4 /*yield*/, this.client.promise];
                    case 1:
                        (_a.sent()).unsubscribe(MatrixClientEvent_1.MatrixClientEventType.MESSAGE, listener);
                        this.activeListeners.delete(senderPublicKey);
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.unsubscribeFromEncryptedMessages = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ;
                        return [4 /*yield*/, this.client.promise];
                    case 1:
                        (_a.sent()).unsubscribeAll(MatrixClientEvent_1.MatrixClientEventType.MESSAGE);
                        this.activeListeners.clear();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.sendMessage = function (message, peer) {
        return __awaiter(this, void 0, void 0, function () {
            var sharedTx, recipientHash, recipient, roomId, encryptedMessage;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createCryptoBoxClient(peer.publicKey, this.keyPair.privateKey)];
                    case 1:
                        sharedTx = (_a.sent()).sharedTx;
                        return [4 /*yield*/, crypto_1.getHexHash(Buffer.from(peer.publicKey, 'hex'))];
                    case 2:
                        recipientHash = _a.sent();
                        recipient = crypto_1.recipientString(recipientHash, peer.relayServer);
                        return [4 /*yield*/, this.getRelevantRoom(recipient)
                            // Before we send the message, we have to wait for the join to be accepted.
                        ];
                    case 3:
                        roomId = _a.sent();
                        // Before we send the message, we have to wait for the join to be accepted.
                        return [4 /*yield*/, this.waitForJoin(roomId)]; // TODO: This can probably be removed because we are now waiting inside the get room method
                    case 4:
                        // Before we send the message, we have to wait for the join to be accepted.
                        _a.sent(); // TODO: This can probably be removed because we are now waiting inside the get room method
                        return [4 /*yield*/, crypto_1.encryptCryptoboxPayload(message, sharedTx)
                            // logger.log(
                            //   'sendMessage',
                            //   'sending encrypted message',
                            //   peer.publicKey,
                            //   roomId,
                            //   message,
                            //   await new Serializer().deserialize(message)
                            // )
                        ];
                    case 5:
                        encryptedMessage = _a.sent();
                        return [4 /*yield*/, this.client.promise];
                    case 6:
                        (_a.sent()).sendTextMessage(roomId, encryptedMessage).catch(function (error) { return __awaiter(_this, void 0, void 0, function () {
                            var newRoomId;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!(error.errcode === 'M_FORBIDDEN')) return [3 /*break*/, 4];
                                        // Room doesn't exist
                                        logger.log("sendMessage", "M_FORBIDDEN", error);
                                        return [4 /*yield*/, this.deleteRoomIdFromRooms(roomId)];
                                    case 1:
                                        _a.sent();
                                        return [4 /*yield*/, this.getRelevantRoom(recipient)];
                                    case 2:
                                        newRoomId = _a.sent();
                                        return [4 /*yield*/, this.client.promise];
                                    case 3:
                                        (_a.sent())
                                            .sendTextMessage(newRoomId, encryptedMessage)
                                            .catch(function (error2) { return __awaiter(_this, void 0, void 0, function () {
                                            return __generator(this, function (_a) {
                                                logger.log("sendMessage", "inner error", error2);
                                                return [2 /*return*/];
                                            });
                                        }); });
                                        return [3 /*break*/, 5];
                                    case 4:
                                        logger.log("sendMessage", "not forbidden", error);
                                        _a.label = 5;
                                    case 5: return [2 /*return*/];
                                }
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.updatePeerRoom = function (sender, roomId) {
        return __awaiter(this, void 0, void 0, function () {
            var split, roomIds, room;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        split = sender.split(':');
                        if (split.length < 2 || !split[0].startsWith('@')) {
                            throw new Error('Invalid sender');
                        }
                        return [4 /*yield*/, this.storage.get(__1.StorageKey.MATRIX_PEER_ROOM_IDS)];
                    case 1:
                        roomIds = _a.sent();
                        room = roomIds[sender];
                        if (room && room[1]) {
                            // If we have a room already, let's ignore it. We need to do this, otherwise it will be loaded from the matrix cache.
                            this.ignoredRooms.push(room[1]);
                        }
                        roomIds[sender] = roomId;
                        return [4 /*yield*/, this.storage.set(__1.StorageKey.MATRIX_PEER_ROOM_IDS, roomIds)
                            // TODO: We also need to delete the room from the sync state
                            // If we need to delete a room, we can assume the local state is not up to date anymore, so we can reset the state
                        ];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.deleteRoomIdFromRooms = function (roomId) {
        return __awaiter(this, void 0, void 0, function () {
            var roomIds, newRoomIds;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.get(__1.StorageKey.MATRIX_PEER_ROOM_IDS)];
                    case 1:
                        roomIds = _a.sent();
                        newRoomIds = Object.entries(roomIds)
                            .filter(function (entry) { return entry[1] !== roomId; })
                            .reduce(function (pv, cv) {
                            var _a;
                            return (__assign(__assign({}, pv), (_a = {}, _a[cv[0]] = cv[1], _a)));
                        }, {});
                        return [4 /*yield*/, this.storage.set(__1.StorageKey.MATRIX_PEER_ROOM_IDS, newRoomIds)
                            // TODO: We also need to delete the room from the sync state
                            // If we need to delete a room, we can assume the local state is not up to date anymore, so we can reset the state
                        ];
                    case 2:
                        _a.sent();
                        // TODO: We also need to delete the room from the sync state
                        // If we need to delete a room, we can assume the local state is not up to date anymore, so we can reset the state
                        this.ignoredRooms.push(roomId);
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.listenForChannelOpening = function (messageCallback) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ;
                        return [4 /*yield*/, this.client.promise];
                    case 1:
                        (_a.sent()).subscribe(MatrixClientEvent_1.MatrixClientEventType.MESSAGE, function (event) { return __awaiter(_this, void 0, void 0, function () {
                            var _a, splits, payload, pairingResponse, _b, _c, _d, _e, _f, decryptionError_2;
                            return __generator(this, function (_g) {
                                switch (_g.label) {
                                    case 0:
                                        _a = this.isTextMessage(event.content);
                                        if (!_a) return [3 /*break*/, 2];
                                        return [4 /*yield*/, this.isChannelOpenMessage(event.content)];
                                    case 1:
                                        _a = (_g.sent());
                                        _g.label = 2;
                                    case 2:
                                        if (!_a) return [3 /*break*/, 9];
                                        logger.log("listenForChannelOpening", "channel opening", JSON.stringify(event));
                                        return [4 /*yield*/, this.updateRelayServer(event.content.message.sender)];
                                    case 3:
                                        _g.sent();
                                        return [4 /*yield*/, this.updatePeerRoom(event.content.message.sender, event.content.roomId)];
                                    case 4:
                                        _g.sent();
                                        splits = event.content.message.content.split(':');
                                        payload = Buffer.from(splits[splits.length - 1], 'hex');
                                        if (!(payload.length >=
                                            sodium.crypto_secretbox_NONCEBYTES + sodium.crypto_secretbox_MACBYTES)) return [3 /*break*/, 9];
                                        _g.label = 5;
                                    case 5:
                                        _g.trys.push([5, 8, , 9]);
                                        _c = (_b = JSON).parse;
                                        return [4 /*yield*/, crypto_1.openCryptobox(payload, this.keyPair.publicKey, this.keyPair.privateKey)];
                                    case 6:
                                        pairingResponse = _c.apply(_b, [_g.sent()]);
                                        _d = messageCallback;
                                        _e = [__assign({}, pairingResponse)];
                                        _f = {};
                                        return [4 /*yield*/, get_sender_id_1.getSenderId(pairingResponse.publicKey)];
                                    case 7:
                                        _d.apply(void 0, [__assign.apply(void 0, _e.concat([(_f.senderId = _g.sent(), _f)]))]);
                                        return [3 /*break*/, 9];
                                    case 8:
                                        decryptionError_2 = _g.sent();
                                        return [3 /*break*/, 9];
                                    case 9: return [2 /*return*/];
                                }
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.waitForJoin = function (roomId, retry) {
        if (retry === void 0) { retry = 0; }
        return __awaiter(this, void 0, void 0, function () {
            var room;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.promise];
                    case 1: return [4 /*yield*/, (_a.sent()).getRoomById(roomId)];
                    case 2:
                        room = _a.sent();
                        logger.log("waitForJoin", "Currently " + room.members.length + " members, we need at least 2");
                        if (room.members.length >= 2 || room.members.length === 0) {
                            // 0 means it's an unknown room, we don't need to wait
                            return [2 /*return*/];
                        }
                        else {
                            if (retry <= 200) {
                                // On mobile, due to app switching, we potentially have to wait for a long time
                                logger.log("Waiting for join... Try: " + retry);
                                return [2 /*return*/, new Promise(function (resolve) {
                                        setTimeout(function () { return __awaiter(_this, void 0, void 0, function () {
                                            return __generator(this, function (_a) {
                                                resolve(this.waitForJoin(roomId, retry + 1));
                                                return [2 /*return*/];
                                            });
                                        }); }, 100 * (retry > 50 ? 10 : 1)); // After the initial 5 seconds, retry only once per second
                                    })];
                            }
                            else {
                                throw new Error("No one joined after " + retry + " tries.");
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.sendPairingResponse = function (pairingRequest) {
        return __awaiter(this, void 0, void 0, function () {
            var recipientHash, recipient, roomId, message, _a, _b, _c, encryptedMessage, msg;
            var _this = this;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        logger.log("sendPairingResponse");
                        return [4 /*yield*/, crypto_1.getHexHash(Buffer.from(pairingRequest.publicKey, 'hex'))];
                    case 1:
                        recipientHash = _d.sent();
                        recipient = crypto_1.recipientString(recipientHash, pairingRequest.relayServer);
                        return [4 /*yield*/, this.client.promise];
                    case 2: return [4 /*yield*/, (_d.sent()).createTrustedPrivateRoom(recipient)];
                    case 3:
                        roomId = _d.sent();
                        return [4 /*yield*/, this.updatePeerRoom(recipient, roomId)
                            // Before we send the message, we have to wait for the join to be accepted.
                        ];
                    case 4:
                        _d.sent();
                        // Before we send the message, we have to wait for the join to be accepted.
                        return [4 /*yield*/, this.waitForJoin(roomId)
                            // TODO: remove v1 backwards-compatibility
                        ]; // TODO: This can probably be removed because we are now waiting inside the get room method
                    case 5:
                        // Before we send the message, we have to wait for the join to be accepted.
                        _d.sent(); // TODO: This can probably be removed because we are now waiting inside the get room method
                        if (!(typeof pairingRequest.version === 'undefined')) return [3 /*break*/, 7];
                        return [4 /*yield*/, this.getPublicKey()]; // v1
                    case 6:
                        _a = _d.sent(); // v1
                        return [3 /*break*/, 9];
                    case 7:
                        _c = (_b = JSON).stringify;
                        return [4 /*yield*/, this.getPairingResponseInfo(pairingRequest)];
                    case 8:
                        _a = _c.apply(_b, [_d.sent()]); // v2
                        _d.label = 9;
                    case 9:
                        message = _a;
                        return [4 /*yield*/, this.encryptMessageAsymmetric(pairingRequest.publicKey, message)];
                    case 10:
                        encryptedMessage = _d.sent();
                        msg = ['@channel-open', recipient, encryptedMessage].join(':');
                        return [4 /*yield*/, this.client.promise];
                    case 11:
                        (_d.sent()).sendTextMessage(roomId, msg).catch(function (error) { return __awaiter(_this, void 0, void 0, function () {
                            var newRoomId;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (!(error.errcode === 'M_FORBIDDEN')) return [3 /*break*/, 4];
                                        // Room doesn't exist
                                        logger.log("sendMessage", "M_FORBIDDEN", error);
                                        return [4 /*yield*/, this.deleteRoomIdFromRooms(roomId)];
                                    case 1:
                                        _a.sent();
                                        return [4 /*yield*/, this.getRelevantRoom(recipient)];
                                    case 2:
                                        newRoomId = _a.sent();
                                        return [4 /*yield*/, this.client.promise];
                                    case 3:
                                        (_a.sent()).sendTextMessage(newRoomId, msg).catch(function (error2) { return __awaiter(_this, void 0, void 0, function () {
                                            return __generator(this, function (_a) {
                                                logger.log("sendMessage", "inner error", error2);
                                                return [2 /*return*/];
                                            });
                                        }); });
                                        return [3 /*break*/, 5];
                                    case 4:
                                        logger.log("sendMessage", "not forbidden", error);
                                        _a.label = 5;
                                    case 5: return [2 /*return*/];
                                }
                            });
                        }); });
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.isTextMessage = function (content) {
        return content.message.type === MatrixMessage_1.MatrixMessageType.TEXT;
    };
    P2PCommunicationClient.prototype.updateRelayServer = function (sender) {
        return __awaiter(this, void 0, void 0, function () {
            var split, senderHash, relayServer, manager, peers, promiseArray;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        split = sender.split(':');
                        if (split.length < 2 || !split[0].startsWith('@')) {
                            throw new Error('Invalid sender');
                        }
                        senderHash = split.shift();
                        relayServer = split.join(':');
                        manager = localStorage.getItem('beacon:communication-peers-dapp')
                            ? new __1.PeerManager(this.storage, __1.StorageKey.TRANSPORT_P2P_PEERS_DAPP)
                            : new __1.PeerManager(this.storage, __1.StorageKey.TRANSPORT_P2P_PEERS_WALLET);
                        return [4 /*yield*/, manager.getPeers()];
                    case 1:
                        peers = _a.sent();
                        promiseArray = peers.map(function (peer) { return __awaiter(_this, void 0, void 0, function () {
                            var hash, _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        _a = "@";
                                        return [4 /*yield*/, crypto_1.getHexHash(Buffer.from(peer.publicKey, 'hex'))];
                                    case 1:
                                        hash = _a + (_b.sent());
                                        if (!(hash === senderHash)) return [3 /*break*/, 3];
                                        if (!(peer.relayServer !== relayServer)) return [3 /*break*/, 3];
                                        peer.relayServer = relayServer;
                                        return [4 /*yield*/, manager.addPeer(peer)];
                                    case 2:
                                        _b.sent();
                                        _b.label = 3;
                                    case 3: return [2 /*return*/];
                                }
                            });
                        }); });
                        return [4 /*yield*/, Promise.all(promiseArray)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.isChannelOpenMessage = function (content) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c, _d, _e, _f;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        _b = (_a = content.message.content).startsWith;
                        _c = "@channel-open:@";
                        _d = crypto_1.getHexHash;
                        _f = (_e = Buffer).from;
                        return [4 /*yield*/, this.getPublicKey()];
                    case 1: return [4 /*yield*/, _d.apply(void 0, [_f.apply(_e, [_g.sent(), 'hex'])])];
                    case 2: return [2 /*return*/, _b.apply(_a, [_c + (_g.sent())])];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.isSender = function (event, senderPublicKey) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _b = (_a = event.content.message.sender).startsWith;
                        _c = "@";
                        return [4 /*yield*/, crypto_1.getHexHash(Buffer.from(senderPublicKey, 'hex'))];
                    case 1: return [2 /*return*/, _b.apply(_a, [_c + (_d.sent())])];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.getRelevantRoom = function (recipient) {
        return __awaiter(this, void 0, void 0, function () {
            var roomIds, roomId, room;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.storage.get(__1.StorageKey.MATRIX_PEER_ROOM_IDS)];
                    case 1:
                        roomIds = _a.sent();
                        roomId = roomIds[recipient];
                        if (!!roomId) return [3 /*break*/, 4];
                        logger.log("getRelevantRoom", "No room found for peer " + recipient + ", checking joined ones.");
                        return [4 /*yield*/, this.getRelevantJoinedRoom(recipient)];
                    case 2:
                        room = _a.sent();
                        roomId = room.id;
                        roomIds[recipient] = room.id;
                        return [4 /*yield*/, this.storage.set(__1.StorageKey.MATRIX_PEER_ROOM_IDS, roomIds)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        logger.log("getRelevantRoom", "Using room " + roomId);
                        return [2 /*return*/, roomId];
                }
            });
        });
    };
    P2PCommunicationClient.prototype.getRelevantJoinedRoom = function (recipient) {
        return __awaiter(this, void 0, void 0, function () {
            var joinedRooms, relevantRooms, room, roomId;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.promise];
                    case 1: return [4 /*yield*/, (_a.sent()).joinedRooms];
                    case 2:
                        joinedRooms = _a.sent();
                        logger.log('checking joined rooms', joinedRooms, recipient);
                        relevantRooms = joinedRooms
                            .filter(function (roomElement) { return !_this.ignoredRooms.some(function (id) { return roomElement.id === id; }); })
                            .filter(function (roomElement) {
                            return roomElement.members.some(function (member) { return member === recipient; });
                        });
                        if (!(relevantRooms.length === 0 || this.ignoredRooms.length > 0)) return [3 /*break*/, 8];
                        logger.log("getRelevantJoinedRoom", "no relevant rooms found, creating new one");
                        return [4 /*yield*/, this.client.promise];
                    case 3: return [4 /*yield*/, (_a.sent()).createTrustedPrivateRoom(recipient)];
                    case 4:
                        roomId = _a.sent();
                        return [4 /*yield*/, this.client.promise];
                    case 5: return [4 /*yield*/, (_a.sent()).getRoomById(roomId)];
                    case 6:
                        room = _a.sent();
                        logger.log("getRelevantJoinedRoom", "waiting for other party to join room: " + room.id);
                        return [4 /*yield*/, this.waitForJoin(roomId)];
                    case 7:
                        _a.sent();
                        logger.log("getRelevantJoinedRoom", "new room created and peer invited: " + room.id);
                        return [3 /*break*/, 9];
                    case 8:
                        room = relevantRooms[0];
                        logger.log("getRelevantJoinedRoom", "channel already open, reusing room " + room.id);
                        _a.label = 9;
                    case 9: return [2 /*return*/, room];
                }
            });
        });
    };
    return P2PCommunicationClient;
}(CommunicationClient_1.CommunicationClient));
exports.P2PCommunicationClient = P2PCommunicationClient;
