// @airgap/beacon-sdk/dist/cjs/ui/alert/Alert.js
"use strict";
// Taken from https://github.com/WalletConnect/walletconnect-monorepo/blob/master/packages/qrcode-modal/src/browser.ts
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openAlert = exports.closeAlerts = exports.closeAlert = void 0;
var colorMode_1 = require("../../colorMode");
var generate_uuid_1 = require("../../utils/generate-uuid");
var replace_in_template_1 = require("../../utils/replace-in-template");
var alert_templates_1 = require("./alert-templates");
var PairingAlert_1 = require("./PairingAlert");
var lastFocusedElement;
var document;
if (typeof window !== 'undefined' && typeof window.document !== 'undefined') {
    document = window.document;
}
var timeout = {};
var addQR = function (dataString) {
    if (typeof dataString === 'string') {
        return "<div id=\"beacon--qr__container\"><div id=\"beacon--qr__copy__container\"><button class=\"beacon-modal__button--outline\" id=\"beacon--qr__copy\">Copy</button></div></div>" + dataString;
    }
    return '';
};
var formatAlert = function (id, body, title, buttons, hasPairingPayload) {
    var callToAction = title;
    var buttonsHtml = buttons.map(function (button, index) {
        return "<button id=\"beacon-alert-" + id + "-" + index + "\" class=\"beacon-modal__button" + (button.style === 'outline' ? '--outline' : '') + "\">" + button.text + "</button>";
    });
    var allStyles = alert_templates_1.alertTemplates.default.css;
    if (hasPairingPayload) {
        allStyles += alert_templates_1.alertTemplates.pair.css;
    }
    var alertContainer = alert_templates_1.alertTemplates.container;
    alertContainer = replace_in_template_1.replaceInTemplate(alertContainer, 'main', hasPairingPayload ? alert_templates_1.alertTemplates.pair.html : alert_templates_1.alertTemplates.default.html);
    alertContainer = replace_in_template_1.replaceInTemplate(alertContainer, 'callToAction', callToAction);
    alertContainer = replace_in_template_1.replaceInTemplate(alertContainer, 'buttons', buttonsHtml.join(' '));
    alertContainer = replace_in_template_1.replaceInTemplate(alertContainer, 'body', body);
    alertContainer = replace_in_template_1.replaceInTemplate(alertContainer, 'id', id);
    if (alertContainer.indexOf('{{') >= 0) {
        var start = alertContainer.indexOf('{{');
        var end = alertContainer.indexOf('}}');
        console.error('Not all placeholders replaced!', alertContainer.substr(start, end - start));
        throw new Error('Not all placeholders replaced!');
    }
    return {
        style: allStyles,
        html: alertContainer
    };
};
/**
 * Close an alert by ID
 *
 * @param id ID of alert
 */
var closeAlert = function (id) {
    return new Promise(function (resolve) {
        var _a;
        var wrapper = document.getElementById("beacon-alert-wrapper-" + id);
        if (!wrapper) {
            return resolve();
        }
        var elm = (_a = wrapper.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById("beacon-alert-modal-" + id);
        if (elm) {
            var animationDuration = 300;
            var localTimeout = timeout[id];
            if (localTimeout) {
                clearTimeout(localTimeout);
                timeout[id] = undefined;
            }
            elm.className = elm.className.replace('fadeIn', 'fadeOut');
            window.setTimeout(function () {
                var parent = wrapper.parentNode;
                if (parent) {
                    parent.removeChild(wrapper);
                }
                if (lastFocusedElement) {
                    ;
                    lastFocusedElement.focus(); // set focus back to last focussed element
                }
                resolve();
            }, animationDuration);
        }
        else {
            resolve();
        }
    });
};
exports.closeAlert = closeAlert;
/**
 * Close all alerts
 */
var closeAlerts = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, new Promise(function (resolve) { return __awaiter(void 0, void 0, void 0, function () {
                var openAlertElements, alertIds_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            openAlertElements = document.querySelectorAll('[id^="beacon-alert-wrapper-"]');
                            if (!(openAlertElements.length > 0)) return [3 /*break*/, 2];
                            alertIds_1 = [];
                            openAlertElements.forEach(function (element) { return __awaiter(void 0, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    alertIds_1.push(element.id.split('-')[3]);
                                    return [2 /*return*/];
                                });
                            }); });
                            return [4 /*yield*/, Promise.all(alertIds_1.map(closeAlert))];
                        case 1:
                            _a.sent();
                            resolve();
                            return [3 /*break*/, 3];
                        case 2:
                            resolve();
                            _a.label = 3;
                        case 3: return [2 /*return*/];
                    }
                });
            }); })
            /**
             * Show an alert
             *
             * @param alertConfig The configuration of the alert
             */
            // eslint-disable-next-line complexity
        ];
    });
}); };
exports.closeAlerts = closeAlerts;
/**
 * Show an alert
 *
 * @param alertConfig The configuration of the alert
 */
// eslint-disable-next-line complexity
var openAlert = function (alertConfig) { return __awaiter(void 0, void 0, void 0, function () {
    var body, title, timer, pairingPayload, disclaimer, closeButtonCallback, id, shadowRootEl, shadowRoot, wrapper, buttons, formattedBody, _a, style, html, styleEl, closeButton, closeButtonClick, disclaimerContainer, colorMode, elm, modal;
    var _b, _c;
    return __generator(this, function (_d) {
        switch (_d.label) {
            case 0:
                body = alertConfig.body;
                title = alertConfig.title;
                timer = alertConfig.timer;
                pairingPayload = alertConfig.pairingPayload;
                disclaimer = alertConfig.disclaimerText;
                closeButtonCallback = alertConfig.closeButtonCallback;
                return [4 /*yield*/, closeAlerts()];
            case 1:
                _d.sent();
                return [4 /*yield*/, generate_uuid_1.generateGUID()];
            case 2:
                id = (_d.sent()).split('-').join('');
                shadowRootEl = document.createElement('div');
                shadowRootEl.setAttribute('id', "beacon-alert-wrapper-" + id);
                shadowRoot = shadowRootEl.attachShadow({ mode: 'open' });
                wrapper = document.createElement('div');
                wrapper.setAttribute('tabindex', "0"); // Make modal focussable
                shadowRoot.appendChild(wrapper);
                buttons = __spreadArrays(((_c = (_b = alertConfig.buttons) === null || _b === void 0 ? void 0 : _b.map(function (button) {
                    var _a, _b;
                    return ({
                        text: button.text,
                        // eslint-disable-next-line @typescript-eslint/unbound-method
                        actionCallback: (_a = button.actionCallback) !== null && _a !== void 0 ? _a : (function () { return Promise.resolve(); }),
                        style: (_b = button.style) !== null && _b !== void 0 ? _b : 'outline'
                    });
                })) !== null && _c !== void 0 ? _c : []));
                formattedBody = pairingPayload ? addQR(body) : body !== null && body !== void 0 ? body : '';
                _a = formatAlert(id, formattedBody, title, buttons, !!(pairingPayload === null || pairingPayload === void 0 ? void 0 : pairingPayload.p2pSyncCode)), style = _a.style, html = _a.html;
                wrapper.innerHTML = html;
                styleEl = document.createElement('style');
                styleEl.textContent = style;
                shadowRoot.appendChild(styleEl);
                if (timer) {
                    timeout[id] = window.setTimeout(function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, closeAlert(id)];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); }, timer);
                }
                document.body.prepend(shadowRootEl);
                closeButton = shadowRoot.getElementById("beacon-alert-" + id + "-close");
                closeButtonClick = function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (closeButtonCallback) {
                                    closeButtonCallback();
                                }
                                return [4 /*yield*/, closeAlert(id)];
                            case 1:
                                _a.sent();
                                return [2 /*return*/];
                        }
                    });
                }); };
                if (disclaimer) {
                    disclaimerContainer = shadowRoot.getElementById("beacon--disclaimer");
                    if (disclaimerContainer) {
                        disclaimerContainer.innerHTML = disclaimer;
                    }
                }
                colorMode = colorMode_1.getColorMode();
                elm = shadowRoot.getElementById("beacon-alert-modal-" + id);
                if (elm) {
                    elm.classList.add("theme__" + colorMode);
                    elm.addEventListener('click', closeButtonClick); // Backdrop click dismisses alert
                }
                modal = shadowRoot.querySelectorAll('.beacon-modal__wrapper');
                if (modal.length > 0) {
                    modal[0].addEventListener('click', function (event) {
                        event.stopPropagation();
                    });
                }
                lastFocusedElement = document.activeElement; // Store which element has been focussed before the alert is shown
                wrapper.focus(); // Focus alert for accessibility
                buttons.forEach(function (button, index) {
                    var buttonElement = shadowRoot.getElementById("beacon-alert-" + id + "-" + index);
                    if (buttonElement) {
                        buttonElement.addEventListener('click', function () { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, closeAlert(id)];
                                    case 1:
                                        _a.sent();
                                        if (!button.actionCallback) return [3 /*break*/, 3];
                                        return [4 /*yield*/, button.actionCallback()];
                                    case 2:
                                        _a.sent();
                                        _a.label = 3;
                                    case 3: return [2 /*return*/];
                                }
                            });
                        }); });
                    }
                });
                if (closeButton) {
                    closeButton.addEventListener('click', function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, closeButtonClick()];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); });
                }
                window.addEventListener('keydown', function (event) { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (!(event.key === 'Escape')) return [3 /*break*/, 2];
                                return [4 /*yield*/, closeButtonClick()];
                            case 1:
                                _a.sent();
                                _a.label = 2;
                            case 2: return [2 /*return*/];
                        }
                    });
                }); });
                if (!pairingPayload) return [3 /*break*/, 4];
                return [4 /*yield*/, PairingAlert_1.preparePairingAlert(shadowRoot, pairingPayload)];
            case 3:
                _d.sent();
                _d.label = 4;
            case 4: return [2 /*return*/, id];
        }
    });
}); };
exports.openAlert = openAlert;
