// @airgap/beacon-sdk/dist/cjs/ui/alert/Pairing.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pairing = exports.WalletType = exports.Platform = void 0;
var Serializer_1 = require("../../Serializer");
var __1 = require("../..");
var MockWindow_1 = require("../../MockWindow");
var get_tzip10_link_1 = require("../../utils/get-tzip10-link");
var platform_1 = require("../../utils/platform");
var wallet_lists_1 = require("./wallet-lists");
var serializer = new Serializer_1.Serializer();
var defaultExtensions = [
    'ookjlbkiijinhpmnjffcofjonbfbgaoc',
    'gpfndedineagiepkpinficbcbbgjoenn' // Beacon
];
var Platform;
(function (Platform) {
    Platform[Platform["DESKTOP"] = 0] = "DESKTOP";
    Platform[Platform["IOS"] = 1] = "IOS";
    Platform[Platform["ANDROID"] = 2] = "ANDROID";
})(Platform = exports.Platform || (exports.Platform = {}));
var WalletType;
(function (WalletType) {
    WalletType["IOS"] = "ios";
    WalletType["ANDROID"] = "android";
    WalletType["EXTENSION"] = "extension";
    WalletType["DESKTOP"] = "desktop";
    WalletType["WEB"] = "web";
})(WalletType = exports.WalletType || (exports.WalletType = {}));
/**
 * @internalapi
 *
 */
var Pairing = /** @class */ (function () {
    function Pairing() {
    }
    Pairing.getPlatfrom = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, platform_1.isAndroid(window) ? Platform.ANDROID : platform_1.isIOS(window) ? Platform.IOS : Platform.DESKTOP];
            });
        });
    };
    Pairing.getPairingInfo = function (pairingPayload, statusUpdateHandler, mobileWalletHandler, platform) {
        return __awaiter(this, void 0, void 0, function () {
            var activePlatform, _a, pairingCode, postmessageSyncCode, preferredNetwork;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(platform !== null && platform !== void 0)) return [3 /*break*/, 1];
                        _a = platform;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, Pairing.getPlatfrom()];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        activePlatform = _a;
                        pairingCode = pairingPayload.p2pSyncCode;
                        postmessageSyncCode = pairingPayload.postmessageSyncCode;
                        preferredNetwork = pairingPayload.preferredNetwork;
                        switch (activePlatform) {
                            case Platform.DESKTOP:
                                return [2 /*return*/, Pairing.getDesktopPairingAlert(pairingCode, statusUpdateHandler, postmessageSyncCode, mobileWalletHandler, preferredNetwork)];
                            case Platform.IOS:
                                return [2 /*return*/, Pairing.getIOSPairingAlert(pairingCode, statusUpdateHandler, preferredNetwork)];
                            case Platform.ANDROID:
                                return [2 /*return*/, Pairing.getAndroidPairingAlert(pairingCode, statusUpdateHandler, preferredNetwork)];
                            default:
                                throw new Error('platform unknown');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Pairing.getDesktopPairingAlert = function (pairingCode, statusUpdateHandler, postmessageSyncCode, mobileWalletHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var availableExtensions, _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4 /*yield*/, __1.availableTransports.availableExtensions];
                    case 1:
                        availableExtensions = _e.sent();
                        availableExtensions.forEach(function (ext) {
                            var index = defaultExtensions.indexOf(ext.id);
                            if (index >= 0) {
                                defaultExtensions.splice(index, 1);
                            }
                        });
                        _a = {};
                        _b = [{
                                title: 'Browser Extensions',
                                type: WalletType.EXTENSION,
                                wallets: __spreadArrays(availableExtensions.map(function (app) {
                                    var _a, _b, _c, _d, _e;
                                    var ext = wallet_lists_1.extensionList.find(function (extEl) { return extEl.id === app.id; });
                                    return {
                                        key: (_a = ext === null || ext === void 0 ? void 0 : ext.key) !== null && _a !== void 0 ? _a : app.id,
                                        name: (_b = app.name) !== null && _b !== void 0 ? _b : ext === null || ext === void 0 ? void 0 : ext.name,
                                        logo: (_c = app.iconUrl) !== null && _c !== void 0 ? _c : ext === null || ext === void 0 ? void 0 : ext.logo,
                                        shortName: (_d = app.shortName) !== null && _d !== void 0 ? _d : ext === null || ext === void 0 ? void 0 : ext.shortName,
                                        color: (_e = app.color) !== null && _e !== void 0 ? _e : ext === null || ext === void 0 ? void 0 : ext.color,
                                        enabled: true,
                                        clickHandler: function () {
                                            return __awaiter(this, void 0, void 0, function () {
                                                var postmessageCode, _a, _b, message;
                                                return __generator(this, function (_c) {
                                                    switch (_c.label) {
                                                        case 0:
                                                            if (!postmessageSyncCode) return [3 /*break*/, 3];
                                                            _b = (_a = serializer).serialize;
                                                            return [4 /*yield*/, postmessageSyncCode()];
                                                        case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                        case 2:
                                                            postmessageCode = _c.sent();
                                                            message = {
                                                                target: __1.ExtensionMessageTarget.EXTENSION,
                                                                payload: postmessageCode,
                                                                targetId: app.id
                                                            };
                                                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                                                            MockWindow_1.windowRef.postMessage(message, MockWindow_1.windowRef.location.origin);
                                                            _c.label = 3;
                                                        case 3:
                                                            statusUpdateHandler(WalletType.EXTENSION, this);
                                                            return [2 /*return*/];
                                                    }
                                                });
                                            });
                                        }
                                    };
                                }), wallet_lists_1.extensionList
                                    .filter(function (app) { return defaultExtensions.some(function (extId) { return extId === app.id; }); })
                                    .map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: false,
                                    clickHandler: function () {
                                        // Don't do anything
                                    }
                                }); })).sort(function (a, b) { return a.key.localeCompare(b.key); })
                            }];
                        _c = {
                            title: 'Desktop & Web Wallets',
                            type: WalletType.DESKTOP
                        };
                        _d = [wallet_lists_1.desktopList.map(function (app) { return ({
                                key: app.key,
                                name: app.name,
                                shortName: app.shortName,
                                color: app.color,
                                logo: app.logo,
                                enabled: true,
                                clickHandler: function () {
                                    return __awaiter(this, void 0, void 0, function () {
                                        var code, _a, _b, link;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    _b = (_a = serializer).serialize;
                                                    return [4 /*yield*/, pairingCode()];
                                                case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                case 2:
                                                    code = _c.sent();
                                                    link = get_tzip10_link_1.getTzip10Link(app.deepLink, code);
                                                    window.open(link, '_blank');
                                                    statusUpdateHandler(WalletType.DESKTOP, this, true);
                                                    return [2 /*return*/];
                                            }
                                        });
                                    });
                                }
                            }); })];
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 2: return [2 /*return*/, (_a.walletLists = _b.concat([
                            (_c.wallets = __spreadArrays.apply(void 0, _d.concat([(_e.sent())])).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _c),
                            {
                                title: 'Mobile Wallets',
                                type: WalletType.IOS,
                                wallets: __spreadArrays(wallet_lists_1.iOSList.map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: true,
                                    clickHandler: function () {
                                        return __awaiter(this, void 0, void 0, function () {
                                            var code, _a, _b;
                                            return __generator(this, function (_c) {
                                                switch (_c.label) {
                                                    case 0:
                                                        _b = (_a = serializer).serialize;
                                                        return [4 /*yield*/, pairingCode()];
                                                    case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                    case 2:
                                                        code = _c.sent();
                                                        mobileWalletHandler(code);
                                                        statusUpdateHandler(WalletType.IOS, this, true);
                                                        return [2 /*return*/];
                                                }
                                            });
                                        });
                                    }
                                }); })).sort(function (a, b) { return a.key.localeCompare(b.key); })
                            }
                        ]),
                            _a.buttons = [],
                            _a)];
                }
            });
        });
    };
    Pairing.getIOSPairingAlert = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _a = {};
                        _b = [{
                                title: 'Mobile Wallets',
                                type: WalletType.IOS,
                                wallets: wallet_lists_1.iOSList
                                    .map(function (app) { return ({
                                    key: app.key,
                                    name: app.name,
                                    shortName: app.shortName,
                                    color: app.color,
                                    logo: app.logo,
                                    enabled: true,
                                    clickHandler: function () {
                                        var _a;
                                        return __awaiter(this, void 0, void 0, function () {
                                            var code, _b, _c, link, a;
                                            return __generator(this, function (_d) {
                                                switch (_d.label) {
                                                    case 0:
                                                        _c = (_b = serializer).serialize;
                                                        return [4 /*yield*/, pairingCode()];
                                                    case 1: return [4 /*yield*/, _c.apply(_b, [_d.sent()])];
                                                    case 2:
                                                        code = _d.sent();
                                                        link = get_tzip10_link_1.getTzip10Link((_a = app.deepLink) !== null && _a !== void 0 ? _a : app.universalLink, code);
                                                        a = document.createElement('a');
                                                        a.setAttribute('href', link);
                                                        a.dispatchEvent(new MouseEvent('click', { view: window, bubbles: true, cancelable: true }));
                                                        statusUpdateHandler(WalletType.IOS, this, true);
                                                        return [2 /*return*/];
                                                }
                                            });
                                        });
                                    }
                                }); })
                                    .sort(function (a, b) { return a.key.localeCompare(b.key); })
                            }];
                        _c = {
                            title: 'Web Wallets',
                            type: WalletType.WEB
                        };
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 1: return [2 /*return*/, (_a.walletLists = _b.concat([
                            (_c.wallets = __spreadArrays.apply(void 0, [(_d.sent())]).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _c)
                        ]),
                            _a.buttons = [],
                            _a)];
                }
            });
        });
    };
    Pairing.getAndroidPairingAlert = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = {};
                        _b = {
                            title: 'Web Wallets',
                            type: WalletType.WEB
                        };
                        return [4 /*yield*/, Pairing.getWebList(pairingCode, statusUpdateHandler, network)];
                    case 1: return [2 /*return*/, (_a.walletLists = [
                            (_b.wallets = __spreadArrays.apply(void 0, [(_c.sent())]).sort(function (a, b) { return a.key.localeCompare(b.key); }),
                                _b)
                        ],
                            _a.buttons = [
                                {
                                    title: 'Mobile Wallets',
                                    text: 'Connect Wallet',
                                    clickHandler: function () { return __awaiter(_this, void 0, void 0, function () {
                                        var code, _a, _b, qrLink;
                                        return __generator(this, function (_c) {
                                            switch (_c.label) {
                                                case 0:
                                                    _b = (_a = serializer).serialize;
                                                    return [4 /*yield*/, pairingCode()];
                                                case 1: return [4 /*yield*/, _b.apply(_a, [_c.sent()])];
                                                case 2:
                                                    code = _c.sent();
                                                    qrLink = get_tzip10_link_1.getTzip10Link('tezos://', code);
                                                    window.open(qrLink, '_blank');
                                                    statusUpdateHandler(WalletType.ANDROID);
                                                    return [2 /*return*/];
                                            }
                                        });
                                    }); }
                                }
                            ],
                            _a)];
                }
            });
        });
    };
    Pairing.getWebList = function (pairingCode, statusUpdateHandler, network) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, wallet_lists_1.webList
                        .map(function (app) { return ({
                        key: app.key,
                        name: app.name,
                        shortName: app.shortName,
                        color: app.color,
                        logo: app.logo,
                        enabled: true,
                        clickHandler: function () {
                            var _a;
                            return __awaiter(this, void 0, void 0, function () {
                                var code, _b, _c, link;
                                return __generator(this, function (_d) {
                                    switch (_d.label) {
                                        case 0:
                                            _c = (_b = serializer).serialize;
                                            return [4 /*yield*/, pairingCode()];
                                        case 1: return [4 /*yield*/, _c.apply(_b, [_d.sent()])];
                                        case 2:
                                            code = _d.sent();
                                            link = get_tzip10_link_1.getTzip10Link((_a = app.links[network]) !== null && _a !== void 0 ? _a : app.links[__1.NetworkType.MAINNET], code);
                                            window.open(link, '_blank');
                                            statusUpdateHandler(WalletType.WEB, this, true);
                                            return [2 /*return*/];
                                    }
                                });
                            });
                        }
                    }); })
                        .sort(function (a, b) { return a.key.localeCompare(b.key); })];
            });
        });
    };
    return Pairing;
}());
exports.Pairing = Pairing;
