// @taquito/taquito/dist/lib/operations/operation-emitter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationEmitter = exports.SIGNATURE_STUB = void 0;
var rpc_1 = require("@taquito/rpc");
var constants_1 = require("../constants");
var operation_errors_1 = require("./operation-errors");
var types_1 = require("./types");
// RPC requires a signature but does not verify it
exports.SIGNATURE_STUB = 'edsigtkpiSSschcaCt9pUVrpNPf7TTcgvgDEDD6NCEHMy8NNQJCGnMfLZzYoQj74yLjo9wx6MPVV29CvVzgi7qEcEUok3k7AuMg';
var OperationEmitter = /** @class */ (function () {
    function OperationEmitter(context) {
        var _this = this;
        this.context = context;
        this.constructOps = function (cOps, publicKeyHash, headCounter, counterFunction, source) {
            var counters = {};
            // tslint:disable strict-type-predicates
            return cOps.map(function (op) {
                switch (op.kind) {
                    case rpc_1.OpKind.ACTIVATION:
                        return __assign({}, op);
                    case rpc_1.OpKind.REVEAL:
                        return __assign(__assign(__assign(__assign({}, op), _this.getSource(op, publicKeyHash, source)), counterFunction(publicKeyHash, headCounter, counters)), _this.getFee(op));
                    case rpc_1.OpKind.ORIGINATION:
                        return __assign(__assign(__assign(__assign(__assign({}, op), { balance: typeof op.balance !== 'undefined' ? "" + op.balance : '0' }), _this.getSource(op, publicKeyHash, source)), counterFunction(publicKeyHash, headCounter, counters)), _this.getFee(op));
                    case rpc_1.OpKind.TRANSACTION:
                        var cops = __assign(__assign(__assign(__assign(__assign({}, op), { amount: typeof op.amount !== 'undefined' ? "" + op.amount : '0' }), _this.getSource(op, publicKeyHash, source)), counterFunction(publicKeyHash, headCounter, counters)), _this.getFee(op));
                        if (cops.source.toLowerCase().startsWith('kt1')) {
                            throw new Error("KT1 addresses are not supported as source since " + constants_1.Protocols.PsBabyM1);
                        }
                        return cops;
                    case rpc_1.OpKind.DELEGATION:
                        return __assign(__assign(__assign(__assign({}, op), _this.getSource(op, publicKeyHash, source)), counterFunction(publicKeyHash, headCounter, counters)), _this.getFee(op));
                    default:
                        throw new Error('Unsupported operation');
                }
            });
        };
        this.getFee = function (op) {
            return {
                // tslint:disable-next-line: strict-type-predicates
                fee: typeof op.fee === 'undefined' ? '0' : "" + op.fee,
                // tslint:disable-next-line: strict-type-predicates
                gas_limit: typeof op.gas_limit === 'undefined' ? '0' : "" + op.gas_limit,
                // tslint:disable-next-line: strict-type-predicates
                storage_limit: typeof op.storage_limit === 'undefined' ? '0' : "" + op.storage_limit,
            };
        };
        this.getSource = function (op, publicKeyHash, source) {
            return {
                source: typeof op.source === 'undefined' ? source || publicKeyHash : op.source,
            };
        };
        this.getContextCounter = function (publicKeyHash, headCounter) {
            if (!_this.context.counters[publicKeyHash] ||
                _this.context.counters[publicKeyHash] < headCounter) {
                _this.context.counters[publicKeyHash] = headCounter;
            }
            var opCounter = ++_this.context.counters[publicKeyHash];
            return {
                counter: "" + opCounter,
            };
        };
        this.getScopeCounter = function (publicKeyHash, headCounter, countersScope) {
            if (!countersScope[publicKeyHash] || countersScope[publicKeyHash] < headCounter) {
                countersScope[publicKeyHash] = headCounter;
            }
            return {
                counter: "" + ++countersScope[publicKeyHash],
            };
        };
    }
    Object.defineProperty(OperationEmitter.prototype, "rpc", {
        get: function () {
            return this.context.rpc;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(OperationEmitter.prototype, "signer", {
        get: function () {
            return this.context.signer;
        },
        enumerable: false,
        configurable: true
    });
    OperationEmitter.prototype.isRevealOpNeeded = function (op, pkh) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.isAccountRevealRequired(pkh)];
                    case 1: return [2 /*return*/, !(_a.sent()) || !this.isRevealRequiredForOpType(op)
                            ? false
                            : true];
                }
            });
        });
    };
    OperationEmitter.prototype.isAccountRevealRequired = function (publicKeyHash) {
        return __awaiter(this, void 0, void 0, function () {
            var manager, haveManager;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.rpc.getManagerKey(publicKeyHash)];
                    case 1:
                        manager = _a.sent();
                        haveManager = manager && typeof manager === 'object' ? !!manager.key : !!manager;
                        return [2 /*return*/, !haveManager];
                }
            });
        });
    };
    OperationEmitter.prototype.isRevealRequiredForOpType = function (op) {
        var e_1, _a;
        var opRequireReveal = false;
        try {
            for (var op_1 = __values(op), op_1_1 = op_1.next(); !op_1_1.done; op_1_1 = op_1.next()) {
                var operation = op_1_1.value;
                if (types_1.isOpRequireReveal(operation)) {
                    opRequireReveal = true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (op_1_1 && !op_1_1.done && (_a = op_1.return)) _a.call(op_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return opRequireReveal;
    };
    OperationEmitter.prototype.prepareOpAndSimulation = function (_a) {
        var operation = _a.operation, source = _a.source;
        return __awaiter(this, void 0, void 0, function () {
            var ops, publicKeyHash, _b, counter, hash, protocol, preparedOpSimulation, preparedOp;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        ops = [];
                        if (Array.isArray(operation)) {
                            ops = __spreadArray([], __read(operation));
                        }
                        else {
                            ops = [operation];
                        }
                        return [4 /*yield*/, this.signer.publicKeyHash()];
                    case 1:
                        publicKeyHash = _c.sent();
                        return [4 /*yield*/, this.getCounterHashAndProtocol(ops, publicKeyHash)];
                    case 2:
                        _b = _c.sent(), counter = _b.counter, hash = _b.hash, protocol = _b.protocol;
                        if (!this.context.counters[publicKeyHash] || this.context.counters[publicKeyHash] < counter) {
                            this.context.counters[publicKeyHash] = counter;
                        }
                        // This is a work around to the counter_in_the_future error returned by the RPC when calling the run_operation
                        // The preparedOpSimulation will be used to prevalidate the operation instead of the preparedOp
                        // The counter of an operation in preparedOpSimulation will be the head counter + 1
                        // The counter in preparedOp will be incremented accordingly if sending many operations in a row by keeping it on the context
                        if (counter !== this.context.counters[publicKeyHash]) {
                            preparedOpSimulation = {
                                opOb: {
                                    branch: hash,
                                    contents: this.constructOps(ops, publicKeyHash, counter, this.getScopeCounter, source),
                                },
                                counter: counter,
                            };
                        }
                        preparedOp = {
                            opOb: {
                                branch: hash,
                                contents: this.constructOps(ops, publicKeyHash, counter, this.getContextCounter, source),
                                protocol: protocol,
                            },
                            counter: counter,
                        };
                        return [2 /*return*/, {
                                preparedOp: preparedOp,
                                preparedOpSimulation: preparedOpSimulation,
                            }];
                }
            });
        });
    };
    OperationEmitter.prototype.getCounterHashAndProtocol = function (ops, publicKeyHash) {
        return __awaiter(this, void 0, void 0, function () {
            var blockHeaderPromise, blockMetaPromise, counterPromise, i, counter_1, _a, header, metadata, headCounter, counter;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        blockHeaderPromise = this.rpc.getBlockHeader();
                        blockMetaPromise = this.rpc.getBlockMetadata();
                        counterPromise = Promise.resolve(undefined);
                        i = 0;
                        _b.label = 1;
                    case 1:
                        if (!(i < ops.length)) return [3 /*break*/, 4];
                        if (!(types_1.isOpRequireReveal(ops[i]) || ops[i].kind === 'reveal')) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.rpc.getContract(publicKeyHash)];
                    case 2:
                        counter_1 = (_b.sent()).counter;
                        counterPromise = Promise.resolve(counter_1);
                        return [3 /*break*/, 4];
                    case 3:
                        i++;
                        return [3 /*break*/, 1];
                    case 4: return [4 /*yield*/, Promise.all([
                            blockHeaderPromise,
                            blockMetaPromise,
                            counterPromise,
                        ])];
                    case 5:
                        _a = __read.apply(void 0, [_b.sent(), 3]), header = _a[0], metadata = _a[1], headCounter = _a[2];
                        if (!header) {
                            throw new Error('Unable to fetch latest block header');
                        }
                        if (!metadata) {
                            throw new Error('Unable to fetch latest metadata');
                        }
                        counter = parseInt(headCounter || '0', 10);
                        return [2 /*return*/, {
                                counter: counter,
                                hash: header.hash,
                                protocol: metadata.next_protocol,
                            }];
                }
            });
        });
    };
    OperationEmitter.prototype.prepareOperationEstimation = function (_a) {
        var operation = _a.operation, source = _a.source;
        return __awaiter(this, void 0, void 0, function () {
            var countersScope, ops, publicKeyHash, _b, _c, counter, hash;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        countersScope = {};
                        ops = [];
                        if (Array.isArray(operation)) {
                            ops = __spreadArray([], __read(operation));
                        }
                        else {
                            ops = [operation];
                        }
                        if (!source) return [3 /*break*/, 1];
                        _b = source;
                        return [3 /*break*/, 3];
                    case 1: return [4 /*yield*/, this.signer.publicKeyHash()];
                    case 2:
                        _b = _d.sent();
                        _d.label = 3;
                    case 3:
                        publicKeyHash = _b;
                        return [4 /*yield*/, this.getCounterHashAndProtocol(ops, publicKeyHash)];
                    case 4:
                        _c = _d.sent(), counter = _c.counter, hash = _c.hash;
                        if (!countersScope[publicKeyHash] || countersScope[publicKeyHash] < counter) {
                            countersScope[publicKeyHash] = counter;
                        }
                        return [2 /*return*/, {
                                opOb: {
                                    branch: hash,
                                    contents: this.constructOps(ops, publicKeyHash, counter, this.getScopeCounter, source),
                                },
                                counter: counter,
                            }];
                }
            });
        });
    };
    OperationEmitter.prototype.preValidate = function (prepared, forgedOperation) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b;
            var _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (!prepared.preparedOpSimulation) return [3 /*break*/, 2];
                        // we call runOperation instead of preapply
                        // to avoid having to produce 2 signatures (op with incremented counter and op used for simulation)
                        _b = this.runOperation;
                        _c = {
                            operation: __assign({ signature: exports.SIGNATURE_STUB }, prepared.preparedOpSimulation.opOb)
                        };
                        return [4 /*yield*/, this.rpc.getChainId()];
                    case 1:
                        // we call runOperation instead of preapply
                        // to avoid having to produce 2 signatures (op with incremented counter and op used for simulation)
                        _a = _b.apply(// If we want to inject many operations in a same block
                        this, [(_c.chain_id = _d.sent(),
                                _c)]);
                        return [3 /*break*/, 3];
                    case 2:
                        _a = this.preapplyOperation(forgedOperation.opOb);
                        _d.label = 3;
                    case 3: return [2 /*return*/, _a];
                }
            });
        });
    };
    OperationEmitter.prototype.forgeOperation = function (_a) {
        var _b = _a.opOb, branch = _b.branch, contents = _b.contents;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_c) {
                return [2 /*return*/, this.context.forger.forge({ branch: branch, contents: contents })];
            });
        });
    };
    OperationEmitter.prototype.forge = function (_a) {
        var _b = _a.opOb, branch = _b.branch, contents = _b.contents, protocol = _b.protocol, counter = _a.counter;
        return __awaiter(this, void 0, void 0, function () {
            var forgedBytes;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.context.forger.forge({ branch: branch, contents: contents })];
                    case 1:
                        forgedBytes = _c.sent();
                        return [2 /*return*/, {
                                opbytes: forgedBytes,
                                opOb: {
                                    branch: branch,
                                    contents: contents,
                                    protocol: protocol,
                                },
                                counter: counter,
                            }];
                }
            });
        });
    };
    OperationEmitter.prototype.simulate = function (op) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = {};
                        return [4 /*yield*/, this.rpc.runOperation(op)];
                    case 1: return [2 /*return*/, (_a.opResponse = _b.sent(),
                            _a.op = op,
                            _a.context = this.context.clone(),
                            _a)];
                }
            });
        });
    };
    OperationEmitter.prototype.estimate = function (_a, estimator) {
        var fee = _a.fee, gasLimit = _a.gasLimit, storageLimit = _a.storageLimit, rest = __rest(_a, ["fee", "gasLimit", "storageLimit"]);
        return __awaiter(this, void 0, void 0, function () {
            var calculatedFee, calculatedGas, calculatedStorage, estimation;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        calculatedFee = fee;
                        calculatedGas = gasLimit;
                        calculatedStorage = storageLimit;
                        if (!(fee === undefined || gasLimit === undefined || storageLimit === undefined)) return [3 /*break*/, 2];
                        return [4 /*yield*/, estimator(__assign({ fee: fee, gasLimit: gasLimit, storageLimit: storageLimit }, rest))];
                    case 1:
                        estimation = _b.sent();
                        if (calculatedFee === undefined) {
                            calculatedFee = estimation.suggestedFeeMutez;
                        }
                        if (calculatedGas === undefined) {
                            calculatedGas = estimation.gasLimit;
                        }
                        if (calculatedStorage === undefined) {
                            calculatedStorage = estimation.storageLimit;
                        }
                        _b.label = 2;
                    case 2: return [2 /*return*/, {
                            fee: calculatedFee,
                            gasLimit: calculatedGas,
                            storageLimit: calculatedStorage,
                        }];
                }
            });
        });
    };
    OperationEmitter.prototype.signOperation = function (forgedBytes) {
        return __awaiter(this, void 0, void 0, function () {
            var signed;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.signer.sign(forgedBytes.opbytes, new Uint8Array([3]))];
                    case 1:
                        signed = _a.sent();
                        forgedBytes.opbytes = signed.sbytes;
                        forgedBytes.opOb.signature = signed.prefixSig;
                        return [2 /*return*/, forgedBytes];
                }
            });
        });
    };
    OperationEmitter.prototype.runOperation = function (op) {
        return __awaiter(this, void 0, void 0, function () {
            var opResponse, results, j, errors;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        opResponse = [];
                        return [4 /*yield*/, this.rpc.runOperation(op)];
                    case 1:
                        results = _a.sent();
                        for (j = 0; j < results.contents.length; j++) {
                            opResponse.push(results.contents[j]);
                        }
                        errors = operation_errors_1.flattenErrors(results);
                        if (errors.length) {
                            // @ts-ignore
                            throw new operation_errors_1.TezosOperationError(errors);
                        }
                        return [2 /*return*/, opResponse];
                }
            });
        });
    };
    OperationEmitter.prototype.preapplyOperation = function (opOb) {
        return __awaiter(this, void 0, void 0, function () {
            var opResponse, results, i, j, errors;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        opResponse = [];
                        return [4 /*yield*/, this.rpc.preapplyOperations([opOb])];
                    case 1:
                        results = _a.sent();
                        if (!Array.isArray(results)) {
                            throw new operation_errors_1.TezosPreapplyFailureError(results);
                        }
                        for (i = 0; i < results.length; i++) {
                            for (j = 0; j < results[i].contents.length; j++) {
                                opResponse.push(results[i].contents[j]);
                            }
                        }
                        errors = operation_errors_1.flattenErrors(results);
                        if (errors.length) {
                            // @ts-ignore
                            throw new operation_errors_1.TezosOperationError(errors);
                        }
                        return [2 /*return*/, opResponse];
                }
            });
        });
    };
    OperationEmitter.prototype.injectOperation = function (signedOperationBytes) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.context.injector.inject(signedOperationBytes)];
            });
        });
    };
    return OperationEmitter;
}());
exports.OperationEmitter = OperationEmitter;
