// @airgap/beacon-sdk/dist/cjs/transports/PostMessageTransport.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostMessageTransport = void 0;
var MockWindow_1 = require("../MockWindow");
var Logger_1 = require("../utils/Logger");
var PeerManager_1 = require("../managers/PeerManager");
var exposed_promise_1 = require("../utils/exposed-promise");
var TransportType_1 = require("../types/transport/TransportType");
var ExtensionMessageTarget_1 = require("../types/ExtensionMessageTarget");
var TransportStatus_1 = require("../types/transport/TransportStatus");
var Origin_1 = require("../types/Origin");
var PostMessageClient_1 = require("./clients/PostMessageClient");
var Transport_1 = require("./Transport");
var logger = new Logger_1.Logger('PostMessageTransport');
var extensions;
/**
 * @internalapi
 *
 *
 */
var PostMessageTransport = /** @class */ (function (_super) {
    __extends(PostMessageTransport, _super);
    function PostMessageTransport(name, keyPair, storage, storageKey) {
        var _this = _super.call(this, name, new PostMessageClient_1.PostMessageClient(name, keyPair), new PeerManager_1.PeerManager(storage, storageKey)) || this;
        _this.type = TransportType_1.TransportType.POST_MESSAGE;
        return _this;
    }
    PostMessageTransport.isAvailable = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve) {
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        var fn = function (event) {
                            var data = event.data;
                            if (data && data.payload === 'pong') {
                                resolve(true);
                                MockWindow_1.windowRef.removeEventListener('message', fn);
                            }
                        };
                        MockWindow_1.windowRef.addEventListener('message', fn);
                        var message = {
                            target: ExtensionMessageTarget_1.ExtensionMessageTarget.EXTENSION,
                            payload: 'ping'
                        };
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        MockWindow_1.windowRef.postMessage(message, MockWindow_1.windowRef.location.origin);
                    })];
            });
        });
    };
    PostMessageTransport.getAvailableExtensions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var localExtensions;
            return __generator(this, function (_a) {
                if (extensions) {
                    return [2 /*return*/, extensions.promise];
                }
                extensions = new exposed_promise_1.ExposedPromise();
                localExtensions = [];
                return [2 /*return*/, new Promise(function (resolve) {
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        var fn = function (event) {
                            var data = event.data;
                            var sender = data.sender;
                            if (data && data.payload === 'pong' && sender) {
                                logger.log('getAvailableExtensions', "extension \"" + sender.name + "\" is available", sender);
                                if (!localExtensions.some(function (ext) { return ext.id === sender.id; })) {
                                    localExtensions.push(sender);
                                }
                            }
                        };
                        MockWindow_1.windowRef.addEventListener('message', fn);
                        setTimeout(function () {
                            // TODO: Should we allow extensions to register after the timeout has passed?
                            MockWindow_1.windowRef.removeEventListener('message', fn);
                            if (extensions) {
                                extensions.resolve(localExtensions);
                            }
                            resolve(localExtensions);
                        }, 1000);
                        var message = {
                            target: ExtensionMessageTarget_1.ExtensionMessageTarget.EXTENSION,
                            payload: 'ping'
                        };
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        MockWindow_1.windowRef.postMessage(message, MockWindow_1.windowRef.location.origin);
                    })];
            });
        });
    };
    PostMessageTransport.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var knownPeers, connectionPromises;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        logger.log('connect');
                        if (this._isConnected !== TransportStatus_1.TransportStatus.NOT_CONNECTED) {
                            return [2 /*return*/];
                        }
                        this._isConnected = TransportStatus_1.TransportStatus.CONNECTING;
                        return [4 /*yield*/, this.getPeers()];
                    case 1:
                        knownPeers = _a.sent();
                        if (knownPeers.length > 0) {
                            logger.log('connect', "connecting to " + knownPeers.length + " peers");
                            connectionPromises = knownPeers.map(function (peer) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, this.listen(peer.publicKey)];
                            }); }); });
                            Promise.all(connectionPromises).catch(function (error) { return logger.error('connect', error); });
                        }
                        return [4 /*yield*/, this.startOpenChannelListener()];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, _super.prototype.connect.call(this)];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    PostMessageTransport.prototype.startOpenChannelListener = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    PostMessageTransport.prototype.getPairingRequestInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.client.getPairingRequestInfo()];
            });
        });
    };
    PostMessageTransport.prototype.listen = function (publicKey) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        logger.log('listen', publicKey);
                        return [4 /*yield*/, this.client
                                .listenForEncryptedMessage(publicKey, function (message, context) {
                                var connectionContext = {
                                    origin: Origin_1.Origin.EXTENSION,
                                    id: context.id
                                };
                                _this.notifyListeners(message, connectionContext).catch(function (error) {
                                    throw error;
                                });
                            })
                                .catch(function (error) {
                                throw error;
                            })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return PostMessageTransport;
}(Transport_1.Transport));
exports.PostMessageTransport = PostMessageTransport;
