// @taquito/ledger-signer/dist/lib/utils.js
"use strict";
/*
 * Some code in this file is adapted from sotez
 * Copyright (c) 2018 Andrew Kishino
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractValue = exports.validateResponse = exports.chunkOperation = exports.appendWatermark = exports.compressPublicKey = exports.transformPathToBuffer = void 0;
var MAX_CHUNK_SIZE = 230;
/**
 *
 * @description Convert the path to a buffer that will be used as LC and CDATA in the APDU send to the ledger device (https://github.com/obsidiansystems/ledger-app-tezos/blob/master/APDUs.md)
 *
 * @param path The ledger derivation path (default is "44'/1729'/0'/0'")
 * @returns A buffer where the first element is the length of the path (default is 4), then 3 bytes for each number of the path to which is added 0x8000000
 */
function transformPathToBuffer(path) {
    var result = [];
    var components = path.split('/');
    components.forEach(function (element) {
        var toNumber = parseInt(element, 10);
        if (Number.isNaN(toNumber)) {
            return;
        }
        if (element.length > 1 && element[element.length - 1] === "'") {
            toNumber += 0x80000000;
        }
        result.push(toNumber);
    });
    var buffer = Buffer.alloc(1 + result.length * 4);
    buffer[0] = result.length;
    result.forEach(function (element, index) {
        buffer.writeUInt32BE(element, 1 + 4 * index);
    });
    return buffer;
}
exports.transformPathToBuffer = transformPathToBuffer;
/**
 *
 * @description Converts uncompressed ledger key to standard tezos binary representation
 */
function compressPublicKey(publicKey, curve) {
    if (curve === 0x00) {
        publicKey = publicKey.slice(1);
    }
    else {
        publicKey[0] = 0x02 + (publicKey[64] & 0x01);
        publicKey = publicKey.slice(0, 33);
    }
    return publicKey;
}
exports.compressPublicKey = compressPublicKey;
function appendWatermark(bytes, watermark) {
    var transactionHex = bytes;
    if (typeof watermark !== 'undefined') {
        var hexWatermark = Buffer.from(watermark).toString('hex');
        transactionHex = hexWatermark.concat(bytes);
    }
    return transactionHex;
}
exports.appendWatermark = appendWatermark;
/**
 *
 * @description In order not to exceed the data length allowed by the Ledger device, split the operation into buffers of 230 bytes (max) and add them to the message to send to the Ledger
 * @param messageToSend The message to send to the Ledger device
 * @param operation The operation which will be chunk if its length is over 230 bytes
 * @returns The instruction to send to the Ledger device
 */
function chunkOperation(messageToSend, operation) {
    var offset = 0;
    while (offset !== operation.length) {
        var chunkSize = offset + MAX_CHUNK_SIZE >= operation.length ? operation.length - offset : MAX_CHUNK_SIZE;
        var buff = Buffer.alloc(chunkSize);
        operation.copy(buff, 0, offset, offset + chunkSize);
        messageToSend.push(buff);
        offset += chunkSize;
    }
    return messageToSend;
}
exports.chunkOperation = chunkOperation;
/**
 *
 * @description Verify if the signature returned by the ledger for tz2 and tz3 is valid
 * @param response The signature returned by the Ledger (return from the signWithLedger function)
 * @returns True if valid, false otherwise
 */
function validateResponse(response) {
    var valid = true;
    if (response[0] !== 0x31 && response[0] !== 0x30) {
        valid = false;
    }
    if (response[1] + 4 !== response.length) {
        valid = false;
    }
    if (response[2] !== 0x02) {
        valid = false;
    }
    var rLength = response[3];
    if (response[4 + rLength] !== 0x02) {
        valid = false;
    }
    var idxLengthSVal = 5 + rLength;
    var sLength = response[idxLengthSVal];
    if ((idxLengthSVal + 1 + sLength) + 2 !== response.length) {
        valid = false;
    }
    return valid;
}
exports.validateResponse = validateResponse;
/**
 *
 * @description Extract a part of the response returned by the Ledger
 * @param idxLength The index in the response from the Ledger that corresponds to the length of the part to extract
 * @param response The signature returned by the Ledger (return from the signWithLedger function)
 * @returns An object that contains the extracted buffer, the index where it starts in the response and the length of the extracted part
 */
function extractValue(idxLength, response) {
    var buffer = Buffer.alloc(32);
    buffer.fill(0);
    var length = response[idxLength];
    var idxValueStart = idxLength + 1;
    if (length > 32) {
        idxValueStart += length - 32;
        length = 32;
    }
    response.copy(buffer, 32 - length, idxValueStart, idxValueStart + length);
    return { buffer: buffer, idxValueStart: idxValueStart, length: length };
}
exports.extractValue = extractValue;
